tools/web_features/manifest.py

```diff
@@ -0,0 +1,249 @@
+MANIFEST.json.
+"""
+
+logger = logging.getLogger(__name__)
+
+MANIFEST_FILE_NAME = "WEB_FEATURES_MANIFEST.json"
+
+
+def abs_path(path: str) -> str:
+    return os.path.abspath(os.path.expanduser(path))
+
+def create_parser() -> argparse.ArgumentParser:
+    """
+    Creates an argument parser for the script.
+
+    Returns:
+        argparse.ArgumentParser: The configured argument parser.
+    """
+    parser = argparse.ArgumentParser(
+        description="Maps tests to web features within a repo root."
+    )
+    parser.add_argument("--repo-root", type=str,
+                        help="The WPT directory. Use this "
+                        "option if the script exists outside the repository")
+    parser.add_argument(
+        "--url-base", action="store", default="/",
+        help="Base url to use as the mount point for tests in this manifest.")
+    parser.add_argument(
+        "-p", "--path", type=abs_path, help="Path to manifest file.")
+    return parser
+
+
+def find_all_test_files_in_dir(root_dir: str, rel_dir_path: str, url_base: str) -> List[SourceFile]:
+    """
+    Finds all test files within a given directory.
+
+    Ignores any SourceFiles that are marked as non_test or the type
+    is SupportFile.item_type
+
+    Args:
+        root_dir (str): The root directory of the repository.
+        rel_dir_path (str): The relative path of the directory to search.
+        url_base (str): Base url to use as the mount point for tests in this manifest.
+
+    Returns:
+        List[SourceFile]: A list of SourceFile objects representing the found test files.
+    """
+    rv: List[SourceFile] = []
+    full_dir_path = os.path.join(root_dir, rel_dir_path)
+    for file in os.listdir(full_dir_path):
+        full_path = os.path.join(full_dir_path, file)
+        rel_file_path = os.path.relpath(full_path, root_dir)
+        source_file = SourceFile(root_dir, rel_file_path, url_base)
+        if not source_file.name_is_non_test and source_file.type != SupportFile.item_type:
+            rv.append(source_file)
+    return rv
+
+@dataclass
+class CmdConfig():
+    """
+    Configuration for the command-line options.
+    """
+
+    repo_root: str  # The root directory of the WPT repository
+    url_base: str  # Base URL used when converting file paths to urls
+
+
+def map_tests_to_web_features(
+        cmd_cfg: CmdConfig,
+        rel_dir_path: str,
+        result: WebFeaturesMap,
+        visited_dirs: Set[str] = set(),
+        prev_inherited_features: List[str] = []) -> None:
+    """
+    Recursively maps tests to web features within a directory structure.
+
+    Args:
+        cmd_cfg (CmdConfig): The configuration for the command-line options.
+        rel_dir_path (str): The relative path of the directory to process.
+        result (WebFeaturesMap): The object to store the mapping results.
+        visited_dirs (Set[str], optional): A set of directories that have already been processed. Defaults to set().
```

Comment:
How can we end up visiting the same directory twice? The recursive call is with a subdirectory, and there are no cycles, right?

Reply:
Thanks for bringing this up.

This was my first time really using os.walk. During my experiments I saw that each iteration of os.walk will traverse the filesystem from top to bottom already [1]. And since I'm also calling my function again on the subdirs, the same folder would eventually be called on the next iteration of os.walk.


[1] `for current_dir, sub_dirs, _ in os.walk(os.path.join(cmd_cfg.repo_root, rel_dir_path)):`

[2] 

```
        for sub_dir in sub_dirs:
            map_tests_to_web_features(
                cmd_cfg,
                os.path.join(rel_dir_path, sub_dir),
                result,
                visited_dirs,
                inherited_features
            )
```
As a result, I needed that `visited_dirs` variable.

But I see what you are saying. I'm going to remove the use of os.walk and only do listdir (or something similar). That way I won't need that visited_dirs variable.